File: 2554-maximum-number-of-integers-to-choose-from-a-range-i/2554-maximum-number-of-integers-to-choose-from-a-range-i.py
from typing import List

class Solution:
    def maxCount(self, banned: List[int], n: int, maxSum: int) -> int:
        """
        Finds the maximum number of integers that can be chosen
        from the range [1, n] under the given constraints.

        Parameters:
        banned (List[int]): List of integers that cannot be chosen.
        n (int): Maximum integer in the range [1, n].
        maxSum (int): Maximum allowed sum of chosen integers.

        Returns:
        int: Maximum number of integers that can be chosen.
        """
        banned_set = set(banned)  # Convert banned list to a set for fast lookups
        current_sum = 0
        count = 0
        
        for num in range(1, n + 1):
            if num in banned_set:
                continue  # Skip numbers in the banned list
            if current_sum + num > maxSum:
                break  # Stop if adding the number exceeds maxSum
            current_sum += num
            count += 1
        
        return count

File: 2554-maximum-number-of-integers-to-choose-from-a-range-i/2554-maximum-number-of-integers-to-choose-from-a-range-i.c
#include <stdbool.h>
#include <stdlib.h>

/**
 * @brief Finds the maximum number of integers that can be chosen
 *        from the range [1, n] under the given constraints.
 *
 * @param banned (int*): Array of banned integers.
 * @param bannedSize (int): Size of the banned array.
 * @param n (int): Maximum integer in the range [1, n].
 * @param maxSum (int): Maximum allowed sum of chosen integers.
 * @return int: Maximum number of integers that can be chosen.
 */
int maxCount(int* banned, int bannedSize, int n, int maxSum) {
    // Step 1: Create a hash set for the banned numbers
    bool* bannedSet = (bool*)calloc(n + 1, sizeof(bool));
    for (int i = 0; i < bannedSize; i++) {
        if (banned[i] <= n) {
            bannedSet[banned[i]] = true;
        }
    }

    // Step 2: Iterate over the range [1, n] and calculate the result
    int currentSum = 0, count = 0;
    for (int num = 1; num <= n; num++) {
        if (bannedSet[num]) {
            continue;  // Skip banned numbers
        }
        if (currentSum + num > maxSum) {
            break;  // Stop if adding num exceeds maxSum
        }
        currentSum += num;
        count++;
    }

    // Step 3: Free memory and return the result
    free(bannedSet);
    return count;
}
